datafusion/src/datasource/listing.rs

```diff
@@ -0,0 +1,297 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! A table that uses the `ObjectStore` listing capability
```

Comment:
I think that it is still a "TODO" to use  `ObjectStore` in this module

What do you think about doing that in this PR? The use of `ObjectStore` seems like one of the key outcomes of this restructuring work to me, so I do wonder about postponing it (especially as this proposed PR is additive and doesn't change the existing providers)

Reply:
I think your are right. I didn't want to add `ObjectStore` useage yet to keep the PR simpler, but it definitively makes sense to do it all at once.

Reply:
Interesting turn of event, I just realized that the `LocalFileReader` reader from `ObjectReader` isn't fully implemented yet. 🙃